posts/2019-02-06-algebraic-api-design.md

```diff
@@ -0,0 +1,830 @@
+  Show.show(
+    _.columns
+      .map(col => col.cells.map(_.show).reverse)
+      .transpose
+      .map(_.mkString("[", ",", "]"))
+      .mkString("\n"))
+
+implicit val showGame: Show[SameGameState] = Show.show {
+  case InProgress(board, score) => show"$board\n\nScore: $score (game in progress)"
+  case Finished(board, score)   => show"$board\n\nScore: $score (game finished)"
+}
+
+implicit val showGameState: Show[GameState[Position, SameGameState, Int]] =
+  Show.show(t => show"""
+                        |${t.position}
+                        |Moves: ${t.playedMoves.reverse}
+                        |""".stripMargin)
+```
+
+### Verifying the Game properties
+
+Now that we have defined the interpreter for `Game`, it is time to ensure that `Game` properties are satisfied. We will do this with property-based testing and the library ScalaCheck. ScalaCheck uses a large number of randomly generated test cases to verify that the given properties hold.
+
+```tut:silent
+import org.scalacheck.{Arbitrary, Gen}
+import org.scalacheck.Gen._
+import org.scalatest._
+import org.scalatest.prop.PropertyChecks
+```
+
+To generate test cases, we have to define how to construct the test inputs. ScalaCheck provides numerous combinators that can be composed to create generators for our domain objects. Here are basic implementations of generators for `GameState` and `Postion`:
+
+```tut:silent
+def colEmpty(size: Int): List[CellState] = List.fill(size)(Empty)
+
+def colNonEmpty(size: Int): Gen[List[CellState]] =
+  for {
+    numFilled <- choose(1, size)
+    filled    <- listOfN(numFilled, choose(0, 5).map(c => Filled(Color(c))))
+  } yield filled ++ colEmpty(size - filled.length)
+
+def gameState(min: Int, max: Int): Gen[GameState[Position, SameGameState, Int]] =
+  for {
+    size      <- choose(min, max)
+    numFilled <- choose(0, size)
+    nonEmpty  <- listOfN(numFilled, colNonEmpty(size)).map(_.map(Column(_)))
+    empty     <- listOfN(size - numFilled, const(colEmpty(size))).map(_.map(Column(_)))
+    score     <- Arbitrary.arbitrary[Int]
+  } yield
+    GameState(
+      playedMoves = List.empty[Position],
+      position = SameGame.evaluateGameState(Board(nonEmpty ++ empty), score),
+      score = score
+    )
+
+def move(boardSize: Int): Gen[Position] =
+  for {
+    col <- choose(0, boardSize)
+    row <- choose(0, boardSize)
+  } yield Position(col, row)
+```
+
+Implementing the property checks is now straight forward:
+
+```tut:book
+class GameTests extends PropSpec with PropertyChecks with Matchers {
+  property("simulation is terminal") {
+    forAll(gameState(4, 8)) { gs =>
+      Game.laws
+        .simulationIsTerminal[IO, Position, SameGameState, Int](gs)
+        .unsafeRunSync()
+    }
+  }
+
+  property("legal move modifies game state") {
+    forAll(gameState(4, 8), move(8)) {
+      case (gs, m) =>
+        Game.laws
+          .legalMoveModifiesGameState[IO, Position, SameGameState, Int](gs, m)
```

Comment:
Missing `unsafeRunSync()`?

Reply:
`legalMoveModifiesGameState` returns `Boolean` so it's ok here.